File: homework2/Saytam_Mishra/q11_isPalindrome.java
package homework2.Saytam_Mishra;
// Technique - Doubly linked list forward-backward two-pointer
// Time Complexity - O(n)
// Space Complexity - O(1)

public class q11_isPalindrome {
    static class Node {
        Node next;
        Node prev;
        int val;

        public Node(int val) {
            this.val = val;
            this.next = null;
            this.prev = null;
        }
    }

    public static boolean checkPalindrome(Node head, Node tail) {
        if(head == null) return true;

        while(head != tail && head.prev != tail) {
            if(head.val != tail.val) {
                return false;
            }
            head = head.next;
            tail = tail.prev;
        }

        return true;
    }

    public static void printLinkedList(Node head) {
        if(head == null) {
            System.out.println("");
            return;
        }

        System.out.print(head.val);
        head = head.next;
        while(head != null) {
            System.out.print(" -> " + head.val);
            head = head.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        // Test Case 1
        Node head1 = new Node(9);
        Node tail1 = head1;

        tail1.next = new Node(2);
        tail1.next.prev = tail1;
        tail1 = tail1.next;
        tail1.next = new Node(4);
        tail1.next.prev = tail1;
        tail1 = tail1.next;
        tail1.next = new Node(2);
        tail1.next.prev = tail1;
        tail1 = tail1.next;
        tail1.next = new Node(9);
        tail1.next.prev = tail1;
        tail1 = tail1.next;

        printLinkedList(head1);
        System.out.println("Linked List 1 is Palindrome: " + checkPalindrome(head1, tail1));

        // Test Case 2
        Node head2 = new Node(9);
        Node tail2 = head2;

        tail2.next = new Node(12);
        tail2.next.prev = tail2;
        tail2 = tail2.next;
        tail2.next = new Node(4);
        tail2.next.prev = tail2;
        tail2 = tail2.next;
        tail2.next = new Node(2);
        tail2.next.prev = tail2;
        tail2 = tail2.next;
        tail2.next = new Node(9);
        tail2.next.prev = tail2;
        tail2 = tail2.next;

        printLinkedList(head2);
        System.out.println("Linked List 2 is Palindrome: " + checkPalindrome(head2, tail2));
    }
}

// Time spent - 16m 36s

File: homework2/Vedant_Pahariya/Q11_IsPalindrome.cpp
// Given a doubly linked list, determine if it is a palindrome.

// Technique used: Two-pointer technique (iterative) and Recursion
// Time Complexity: O(n) where n is the number of nodes in the list, it is actually O(n/2) but we drop the constant factor
// Space Complexity: O(1) for iterative approach, O(n) for recursive approach due to recursion stack

#include <bits/stdc++.h>
using namespace std;

struct Node {
  int data;
  Node* next;
  Node* prev;
};

bool ispalinrecur(Node* head, Node* tail){

    if(tail->next == head || tail == head){
        return true;
    }

    bool condition = false;
    if(head->data == tail->data){
        condition = true;
    }

    return ispalinrecur(head->next, tail->prev) && condition;
}

bool ispalin(Node* head, Node* tail){
    while(head != NULL && tail != NULL){
        if(head->data != tail->data){
            return false;
        }
        if(head == tail || head->next == tail){
            break;
        }
        head = head->next;
        tail = tail->prev;
    }
    return true;
}

int main(){
    Node* head = new Node();
    head->data = 1;
    head->prev = NULL;  // Add this
    head->next = new Node();
    
    head->next->data = 2;
    head->next->prev = head;  // Add this
    head->next->next = new Node();
    
    head->next->next->data = 2;
    head->next->next->prev = head->next;  // Add this
    head->next->next->next = new Node();
    
    head->next->next->next->data = 1;
    head->next->next->next->prev = head->next->next;  // Add this
    head->next->next->next->next = NULL;

    Node* tail = head;
    while(tail->next != NULL){
        tail = tail->next;
    }

    cout << ispalinrecur(head, tail) << "\n"; // should return true

    head->next->data = 3; // change second element to break palindrome

    cout << ispalin(head, tail) << "\n"; // should return false

    return 0;
}